src/util/pg.py

```diff
@@ -196,27 +197,56 @@ def explode_query_range(cr, query, table, alias=None, bucket_size=10000, prefix=
+        query += sep_kw + "{parallel_filter}"
+
+    cr.execute(format_query(cr, "SELECT min(id), max(id) FROM {}", table))
     min_id, max_id = cr.fetchone()
     if min_id is None:
         return []  # empty table
+    count = (max_id + 1 - min_id) // bucket_size
+    if count > MAX_BUCKETS:
+        _logger.getChild("explode_query_range").warning(
+            "High number of queries generated (%s); switching to a precise bucketing strategy", count
+        )
+        cr.execute(
+            format_query(
+                cr,
+                """
+                WITH t AS (
+                    SELECT id,
+                           mod(row_number() OVER(ORDER BY id) - 1, %s) AS g
+                      FROM {table}
+                     ORDER BY id
+                ) SELECT array_agg(id ORDER BY id) FILTER (WHERE g=0),
+                         min(id),
+                         max(id)
+                    FROM t
+                """,
+                table=table,
+            ),
+            [bucket_size],
+        )
+        ids, min_id, max_id = cr.fetchone()
+    else:
+        ids = list(range(min_id, max_id + 1, bucket_size))
 
-    if "{parallel_filter}" not in query:
-        sep_kw = " AND " if re.search(r"\sWHERE\s", query, re.M | re.I) else " WHERE "
-        query += sep_kw + "{parallel_filter}"
+    assert min_id == ids[0] and max_id + 1 != ids[-1]  # sanity checks
+    ids.append(max_id + 1)  # ensure last bucket covers whole range
+    # `ids` holds a list of values marking the interval boundaries for all buckets
 
-    if ((max_id - min_id + 1) * 0.9) <= bucket_size:
-        # If there is less than `bucket_size` records (with a 10% tolerance), no need to explode the query.
-        # Force usage of `prefix` in the query to validate it correctness.
-        # If we don't the query may only be valid if there is no split. It avoid scripts to pass the CI but fail in production.
+    if ((max_id - min_id + 1) * 0.9) <= bucket_size or len(ids) == 2:
+        # Do not explode if there are less than `bucket_size` records (with a 10% tolerance), or only one bucket.
+        # Since the query may only be valid if there is no split, force usage of `prefix` in the query to validate its
+        # correctness and avoid scripts that pass the CI but fail in production.
         parallel_filter = "{alias}.id IS NOT NULL".format(alias=alias)
         return [query.format(parallel_filter=parallel_filter)]
 
     parallel_filter = "{alias}.id BETWEEN %(lower-bound)s AND %(upper-bound)s".format(alias=alias)
     query = query.replace("%", "%%").format(parallel_filter=parallel_filter)
     return [
-        cr.mogrify(query, {"lower-bound": index, "upper-bound": index + bucket_size - 1}).decode()
-        for index in range(min_id, max_id, bucket_size)
+        cr.mogrify(query, {"lower-bound": low, "upper-bound": upp - 1}).decode() for (low, upp) in zip(ids, ids[1:])
```

Comment:
Use [itertools.pairwise](https://docs.python.org/3.12/library/itertools.html#itertools.pairwise) instead (we should have a shim in `util.misc` (note: python2 code should use `izip`))
For big buckets, `ids` will be huge, so better not copy it.

Reply:
Well, as much as I like pairwise:
> New in version 3.10.

But we don't need to overdo it -- the good old C-style iteration should be fine here.